### PR TITLE
[x86/Linux] Fix crashes during unwinding with DAC

### PR DESCRIPTION
**Pull request by ayuckhulk**
This PR fixes a couple of crashes when context does not contain all register values. This happens when unwinding is performed through DAC by ICorDebug-based debugger on x86 linux.

@janvorli PTAL

CC @alpencolt

### CODE REVIEW

File: src/unwinder/i386/unwinder_i386.cpp [hunk 1]
```diff
@@ -30,7 +30,7 @@ BOOL OOPStackUnwinderX86::Unwind(T_CONTEXT* pContextRecord, T_KNONVOLATILE_CONTE
     EECodeInfo codeInfo;
     codeInfo.Init((PCODE) ControlPc);
 
-    if (!UnwindStackFrame(&rd, &codeInfo, UpdateAllRegs, &codeManState, NULL))
```
[the remark this answers is not in the document]
Comment (ayuckhulk) [reply src/unwinder/i386/unwinder_i386.cpp]: Stack trace looked like this:
```
003, name = 'netcoredbg', stop reason = signal SIGSEGV: invalid address (fault address: 0x0)
  * frame #0: 0xb6570559 libmscordaccore.so`EECodeInfo::GetSavedMethodCode() [inlined] EECodeInfo::GetStartAddress(this={EECodeInfo}) + 3 at jitinterface.cpp:14003 [opt]
    frame #1: 0xb6570556 libmscordaccore.so`EECodeInfo::GetSavedMethodCode(this={EECodeInfo}) + 22 at jitinterface.cpp:13991 [opt]
    frame #2: 0xb65619c5 libmscordaccore.so`UnwindStackFrame(pContext=<unavailable>, pCodeInfo={EECodeInfo}, flags=<unavailable>, pState=<unavailable>, pUnwindInfo=<unavai
lable>) + 53 at eetwain.cpp:3979 [opt]
    frame #3: 0xb65a734f libmscordaccore.so`OOPStackUnwinderX86::Unwind(pContextRecord=<unavailable>, pContextPointers=<unavailable>) + 303 at unwinder_i386.cpp:33 [opt]
    frame #4: 0xb65a74cb libmscordaccore.so`DacUnwindStackFrame(pContextRecord={CONTEXT}, pContextPointers={KNONVOLATILE_CONTEXT_POINTERS}) + 43 at unwinder_i386.cpp:134 [
opt]
    frame #5: 0xb65883e6 libmscordaccore.so`Thread::VirtualUnwindCallFrame(pContext={CONTEXT}, pContextPointers={KNONVOLATILE_CONTEXT_POINTERS}, pCodeInfo={EECodeInfo}) +
38 at stackwalk.cpp:651 [opt]
    frame #6: 0xb6560c49 libmscordaccore.so`EECodeManager::EnsureCallerContextIsValid(pRD={REGDISPLAY}, pCacheEntry=null, pCodeInfo={EECodeInfo}) + 121 at eetwain.cpp:3092
 [opt]
    frame #7: 0xb658a3bc libmscordaccore.so`StackFrameIterator::CheckForSkippedFrames(this={StackFrameIterator}) + 92 at stackwalk.cpp:3034 [opt]
    frame #8: 0xb658a310 libmscordaccore.so`StackFrameIterator::NextRaw() [inlined] StackFrameIterator::ProcessCurrentFrame(this=<unavailable>) + 221 at stackwalk.cpp:2974
 [opt]
    frame #9: 0xb658a233 libmscordaccore.so`StackFrameIterator::NextRaw(this=<unavailable>) + 1251 at stackwalk.cpp:2741 [opt]
    frame #10: 0xb658890f libmscordaccore.so`StackFrameIterator::Next(this={StackFrameIterator}) + 63 at stackwalk.cpp:1604 [opt]
    frame #11: 0xb64b2615 libmscordaccore.so`DacDbiInterfaceImpl::UnwindStackWalkFrame(this=<unavailable>, pSFIHandle=<unavailable>) + 277 at dacdbiimplstackwalk.cpp:233 [
opt]
    frame #12: 0xb64b28aa libmscordaccore.so`non-virtual thunk to DacDbiInterfaceImpl::UnwindStackWalkFrame(this={DacDbiInterfaceImpl}, pSFIHandle=null) + 42 at dacdbiimpl
stackwalk.cpp:201 [opt]
    frame #13: 0xb67c0a40 libmscordbi.so`CordbStackWalk::Next() [inlined] CordbStackWalk::UnwindStackFrame(this={CordbStackWalk}) + 33 at rsstackwalk.cpp:426 [opt]
    frame #14: 0xb67c0a1f libmscordbi.so`CordbStackWalk::Next(this={CordbStackWalk}) + 559 at rsstackwalk.cpp:478 [opt]
    frame #15: 0xb67c0d2f libmscordbi.so`non-virtual thunk to CordbStackWalk::Next(this={CordbStackWalk}) + 31 at rsstackwalk.cpp:446 [opt]
    frame #16: 0xb675b990 libmscordbi.so`ShimStackWalk::Populate(this=<unavailable>) + 3056 at shimstackwalk.cpp:491 [opt]
    frame #17: 0xb675ad1b libmscordbi.so`ShimStackWalk::ShimStackWalk(this={ShimStackWalk}, pProcess=<unavailable>, pThread={CordbThread}) + 219 at shimstackwalk.cpp:34 [o
pt]
    frame #18: 0xb6752da3 libmscordbi.so`ShimProcess::LookupOrCreateShimStackWalk(this=<unavailable>, pThread=<unavailable>) + 259 at shimprocess.cpp:1636 [opt]
    frame #19: 0xb6722973 libmscordbi.so`CordbThread::GetActiveFrame(this=<unavailable>, ppFrame=<unavailable>) + 403 at rsthread.cpp:1089 [opt]
    frame #20: 0xb6722c48 libmscordbi.so`non-virtual thunk to CordbThread::GetActiveFrame(this={CordbThread}, ppFrame=null) + 40 at rsthread.cpp:1069 [opt]
    frame #21: 0x0806b059 netcoredbg`ManagedCallback::Exception(ICorDebugAppDomain*, ICorDebugThread*, int) + 153
    frame #22: 0xb6757a90 libmscordbi.so`ShimProxyCallback::Exception(this={ExceptionEvent}, args={ManagedEvent::DispatchArgs})::ExceptionEvent::Dispatch(ManagedEvent::Dis
patchArgs) + 64 at shimcallback.cpp:246 [opt]
    frame #23: 0xb679ccc2 libmscordbi.so`CordbProcess::DispatchRCEvent(this=<unavailable>) + 482 at process.cpp:4632 [opt]
    frame #24: 0xb67a7fe8 libmscordbi.so`CordbRCEventThread::FlushQueuedEvents(this=<unavailable>, process=<unavailable>) + 184 at process.cpp:10187 [opt]
    frame #25: 0xb67a8920 libmscordbi.so`CordbRCEventThread::ThreadProc(this=<unavailable>) + 1296 at process.cpp:10442 [opt]
    frame #26: 0xb67a8b0c libmscordbi.so`CordbRCEventThread::ThreadProc(parameter={CordbRCEventThread}) + 28 at process.cpp:10508 [opt]
    frame #27: 0xb65efdbd libmscordaccore.so`CorUnix::CPalThread::ThreadEntry(pvParam={CorUnix::CPalThread}) + 333 at thread.cpp:1749 [opt]
    frame #28: 0xb76c35b6 libpthread.so.0`??? + 214
    frame #29: 0xb74150de libc.so.6`??? + 12
```
This happend during debugging of a huge proprietary C#/C++ application (so native and managed frames are mixed in the debuggee).

[the remark this answers is not in the document]
Comment (janvorli) [reply src/unwinder/i386/unwinder_i386.cpp]: @ayuckhulk could you please take a look to see what was the ControlPC passed to the UnwindStackFrame in this case? Can you check if it is a valid address at all and if it is a managed code address or native code address? As I've said, we should never enter this function with an address that's not in managed code, so if we do, it is a problem somewhere upstream and not in this function.

[the remark this answers is not in the document]
Comment (ayuckhulk) [reply src/unwinder/i386/unwinder_i386.cpp]: @janvorli The `ControlPC` value is `0`.
Please note the the crash happens in the [debugger's](https://github.com/Samsung/netcoredbg/blob/9efdff27b680f6e36fa4397e4fee98925a4ea7e5/src/debug/netcoredbg/manageddebugger.cpp#L358) `ICorDebugManagedCallback::Exception()` callback when calling `ICorDebugThread::GetActiveFrame()`, **not** inside the C# application. Does the assumption about managed code address still holds in such case?

[the remark this answers is not in the document]
Comment (janvorli) [reply src/unwinder/i386/unwinder_i386.cpp]: > Does the assumption about managed code address still holds in such case?

Yes, it does.

**Comment by jkotas** [comment]
Closing stale PRs

**Pull request closed without merging**